BEAM-3026 Adding retrying behavior on ElasticSearchIO for 429: Too Many Requests

aalbatross opened the pull request:
@timrobertson100 

Added Retry behavior in ElasticSearchIO for too many  #requests.
Changes made : 
1. moved RetryConfiguration and RetryPredicate to common place
2. Implemented Retry behavior in ESIO and added tests and verified the change.
3. updated SolrIO, and SolrIOTest to use retry configuration from common place.

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [X] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | --- | --- | --- | ---

aalbatross commented:
retest this please

timrobertson100 commented:
Thanks @aalbatross - I can’t review today but will tomorrow or Wed at latest.

aalbatross commented:
Sure @timrobertson100

timrobertson100 commented on sdks/java/core/src/main/java/org/apache/beam/sdk/io/common/retry/RetryConfiguration.java:
```diff
@@ -0,0 +1,80 @@
+/**
+ * A POJO encapsulating a configuration for retry behavior when issuing requests. A retry will be
+ * attempted until the maxAttempts or maxDuration is exceeded, whichever comes first, for any of the
+ * following exceptions:
```
Incomplete JDoc - don't think it needs to mention the exceptions

timrobertson100 commented on sdks/java/core/src/main/java/org/apache/beam/sdk/io/common/retry/RetryConfiguration.java:
```diff
@@ -0,0 +1,80 @@
+
+  private int maxAttempts;
+  private Duration maxDuration;
+  private RetryPredicate retryPredicate;
```
The original design of `SolrIO` deliberately kept the predicate package private and visible only for tests.
That design was chosen to abide by the policy of Beam IOs which is to only expose tuning for those things that a user might reasonably need to tune. Personally I am not as strict as others, but I do think that in the case of retries the IO should be able to make the decision and this shouldn't be exposed to the user to modify.

echauchot replied:
+1 to that, indeed you described well the no knobs philosophy of Beam

timrobertson100 commented on sdks/java/core/src/main/java/org/apache/beam/sdk/io/common/retry/package-info.java:
```diff
@@ -0,0 +1,20 @@
+ * limitations under the License.
+ */
+
+/** Apache Beam IO Retry mechanism default implementations. */
```
This package doesn't contain default implementations does it? Isn't it just the configuration objects?

timrobertson100 commented on sdks/java/io/elasticsearch-tests/elasticsearch-tests-5/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java:
```diff
@@ -55,7 +68,9 @@
 @ThreadLeakScope(ThreadLeakScope.Scope.NONE)
 public class ElasticsearchIOTest extends ESIntegTestCase implements Serializable {
-
+  //RetryPredicate for 429- TOO MANY REQUESTS.
```
Inline comment not necessary

timrobertson100 commented on sdks/java/io/elasticsearch-tests/elasticsearch-tests-5/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java:
```diff
@@ -216,4 +231,42 @@ public void testWritePartialUpdate() throws Exception {
+
+    Response response =
+        getRestClient().performRequest("POST", "/_bulk", Collections.emptyMap(), entity);
+    //Cannnot create a mock Response, so tried with a bulk request and expecting 400 code.
```
Inline comment not necessary

timrobertson100 commented on sdks/java/io/elasticsearch-tests/elasticsearch-tests-5/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java:
```diff
@@ -216,4 +231,42 @@ public void testWritePartialUpdate() throws Exception {
   }
+
+  @Test
+  public void testDefaultRetryPredicate() throws IOException {
```
Suggest add:

```
/** Test that the default predicate correctly parses chosen error code. */ 
```

timrobertson100 replied:
This test should be in `elasticsearch-tests-common` and implemented for both ES2 and ES5

timrobertson100 commented on sdks/java/io/elasticsearch-tests/elasticsearch-tests-5/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java:
```diff
@@ -216,4 +231,42 @@ public void testWritePartialUpdate() throws Exception {
+        new ElasticsearchIO.DefaultRetryPredicate(400).test(new ResponseException(response)));
+  }
+
+  @Test
```
Suggest add:

```
  /**
   * Test that retries are invoked when Elasticsearch returns a specific error code. We invoke this by issuing 
   * corrupt data and retrying on the `400` error code. Normal behaviour is to retry on `429` only
   * but that is difficult to simulate reliably. The logger is used to verify expected behavior.
   */
```

timrobertson100 commented on sdks/java/io/elasticsearch-tests/elasticsearch-tests-5/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java:
```diff
@@ -216,4 +231,42 @@ public void testWritePartialUpdate() throws Exception {
+    expectedException.expect(IOException.class);
+    expectedException.expectMessage(
+        String.format(ElasticsearchIO.Write.WriteFn.RETRY_FAILED_LOG, 2));
+    // wrong json values to generate 400 http code.
```
Not necessary with the JDoc above

timrobertson100 commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -714,6 +729,57 @@ public void close() throws IOException {
+   * will be attempted until the maxAttempts or maxDuration is exceeded, whichever comes first, for
+   * TOO_MANY_REQUESTS(429) exception.
+   */
+  static class DefaultRetryPredicate implements RetryPredicate {
```
This is package private, but you have modified the configuration to require users to provide this. I expect that won't work outside of the package will it?

See previous comment where I recommend not letting users provide this at all though.

timrobertson100 commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -714,6 +729,57 @@ public void close() throws IOException {
+
+    private int errorCode;
+
+    public DefaultRetryPredicate(@Nonnull int code) {
```
@Nonnull seems redundant for a primitive

timrobertson100 commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -714,6 +729,57 @@ public void close() throws IOException {
+    private int errorCode;
+
+    public DefaultRetryPredicate(@Nonnull int code) {
+      Objects.requireNonNull(code);
```
As above

timrobertson100 commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -714,6 +729,57 @@ public void close() throws IOException {
+    }
+
+    /**
+     * method identified if the error is 429 returns true.
```
Incomplete JDoc

timrobertson100 commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -714,6 +729,57 @@ public void close() throws IOException {
+      try {
+        searchResult = parseResponse(response);
+      } catch (IOException e) {
+        LOG.warn("Could not identify the response " + response);
```
Please always use `LOG.warn("Parameterize messages {}", "like this");`

timrobertson100 commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -714,6 +729,57 @@ public void close() throws IOException {
+     * @param response
+     * @return
+     */
+    private boolean identify(Response response) {
```
Suggest replace this method with:

```
/** Returns true if the response has the error code for any mutation. */
private static boolean errorCodePresent(Response response, int errorCode) {
  try {
    JsonNode json = parseResponse(response);
    if (json.path("errors").asBoolean()) {
      for (JsonNode item : json.path("items")) {
      if (item.findValue("status").asInt() == errorCode) {
        return true;
      }
    }
  } catch (IOException e) {
    LOG.warn("Could not extract error codes from response {}", response);
    return false;
  }
  return false;
}
```

timrobertson100 commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -875,6 +951,15 @@ public PDone expand(PCollection<String> input) {
       private static final int DEFAULT_RETRY_ON_CONFLICT = 5; // race conditions on updates
 
+      private static final Duration RETRY_INITIAL_BACKOFF = Duration.standardSeconds(5);
+      private static final Duration RETRY_MAX_BACKOFF = Duration.standardSeconds(365);
```
365 seconds seems unusual? Days?

timrobertson100 commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -875,6 +951,15 @@ public PDone expand(PCollection<String> input) {
+      private static final Duration RETRY_INITIAL_BACKOFF = Duration.standardSeconds(5);
+      private static final Duration RETRY_MAX_BACKOFF = Duration.standardSeconds(365);
+
+      static final String RETRY_ATTEMPT_LOG = "Retrying again .. %d";
```
Suggest:

```
 @VisibleForTesting
static final String RETRY_ATTEMPT_LOG = "Error writing to Elasticsearch. Retry attempt[%d]";
```

timrobertson100 commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -861,6 +932,11 @@ public Write withUsePartialUpdate(boolean usePartialUpdate) {
       return builder().setUsePartialUpdate(usePartialUpdate).build();
     }
 
+    public Write withRetryConfiguration(RetryConfiguration retryConfiguration) {
```
Needs JDoc (see SolrIO for example)

aalbatross commented:
@timrobertson100 
Thanks for your comments, I have incorporated them. 
I still went with common code approach but now not providing any way for user to set RetryPredicate. The common code is the common retry data structure which could be extended by IO developers to reuse them.

timrobertson100 commented on sdks/java/io/elasticsearch-tests/elasticsearch-tests-5/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java:
```diff
@@ -216,4 +222,38 @@ public void testWritePartialUpdate() throws Exception {
+   * behavior.
+   */
+  @Test
+  public void testWriteRetry() throws Throwable {
```
This test should be in `elasticsearch-tests-common` and tested for both ES2 and ES5

aalbatross commented:
Tim thank you for reviewing this quickly. I have added the changes as you commented.

I am not going for the autovalue/ builder approach now as we spent lot of time doing code change and review using the current approach i do not think it would be great to change the entire approach now and have fresh changes again. 

But if you think it is necessary for any important reasons i will change them.

timrobertson100 commented:
@Thanks Ravi for incorporating the change requests.

I'm -0 on the common configuration (doesn't add much of benefit, spreads otherwise encapsulated code over packages, likely more total lines of code, inconsistent compared to other IO) but I recognise this is subjective and the implementation doesn't expose more configurable parameters than desired.

@jbonofre can you PTAL?

timrobertson100 commented:
@echauchot can you PTAL?

aalbatross commented:
Sure @timrobertson100

timrobertson100 commented on sdks/java/io/elasticsearch-tests/elasticsearch-tests-2/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java:
```diff
@@ -216,8 +216,20 @@ public void testWritePartialUpdate() throws Exception {
   }
 
   @Test
+<<<<<<< HEAD
```
This looks like a bad commit I'm afraid

echauchot commented:
@timrobertson100 @aalbatross starting to take a look. Thanks for the contribution by the way !

echauchot commented on sdks/java/io/elasticsearch-tests/elasticsearch-tests-common/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTestCommon.java:
```diff
@@ -512,4 +527,44 @@ public Void apply(Iterable<String> input) {
+  public void testWriteRetry() throws Throwable {
+    expectedException.expect(IOException.class);
+    expectedException.expectMessage(
+        String.format(ElasticsearchIO.Write.WriteFn.RETRY_FAILED_LOG, 2));
```
why is this 2? 3 -1 ?
please use constants

aalbatross commented:
Thank you @echauchot for your comments, i will incorporate them and come up with AutoValue Builder solution for ESIO.

timrobertson100 commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -893,6 +1013,18 @@ public PDone expand(PCollection<String> input) {
       private static final int DEFAULT_RETRY_ON_CONFLICT = 5; // race conditions on updates
 
+      private static final Duration RETRY_INITIAL_BACKOFF = Duration.standardSeconds(5);
+      private static final Duration RETRY_MAX_BACKOFF = Duration.standardDays(365);
```
Is not needed - see SolrIO:

https://github.com/apache/beam/blob/master/sdks/java/io/solr/src/main/java/org/apache/beam/sdk/io/solr/SolrIO.java#L610

timrobertson100 commented on sdks/java/core/src/main/java/org/apache/beam/sdk/io/common/retry/BaseRetryConfiguration.java:
```diff
@@ -0,0 +1,52 @@
+/*
```
@echauchot 
Moving withRetryPredicate to the BaseRetryConfiguration will expose it in the API. Currently it is package private, visibleForTesting which was by design in SolrIO.

aalbatross replied:
@timrobertson100 @echauchot i agree moving withRetryPredicate to BaseRetryConfiguration will expose API, the only way i can avoid is by overriding the method with protected final key word in ESIO or SolrIO. This will lead to duplicate code in all the implementations which use the common configuration classes.

I think it is better to have duplicates with AutoValue Builder in ESIO and SolrIO instead of using common config objects. I will remove common code.

echauchot replied:
@timrobertson100 you're right about exposing withRetryPredicate, it is better to hide it from the user and keep this knowledge at the IO level.

aalbatross commented on sdks/java/io/elasticsearch-tests/elasticsearch-tests-common/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTestCommon.java:
```diff
@@ -512,4 +527,44 @@ public Void apply(Iterable<String> input) {
+  /**
+   * Test that retries are invoked when Elasticsearch returns a specific error code. We invoke this
+   * by issuing corrupt data and retrying on the `400` error code. Normal behaviour is to retry on
+   * `429` only but that is difficult to simulate reliably. The logger is used to verify expected
```
@echauchot the test is interesting, but i am afraid i can't think of a way to simulate that condition(since the retry code is internal and i cant change the request when the pipeline is executing). Please suggest !

echauchot replied:
Indeed very difficult to simulate with an http 400 bad formatted doc. The batch will be retried until maxAttempts is reached because it will still contain a bad formatted doc. The only way would be to test with an error that can be temporary (like an ES status) and control the embeded ES through a separate thread in the UT to change that status. But now that you changed the logic in `handleRetry()` and considering that you test your predicate in a separate UT, I think it is fine.

aalbatross commented:
@echauchot 
Updated the PR based on comments:

Changes are as follows :
1. Updated the design remove the common classes, used AutoValue Builder for RetryConfiguration In ESIO,
2. RetryPredicate using Response object instead of using Throwable.
3. Removed maxDuration default in ESIO.
4. Updated the retry logic with similar logic provided in comment.
5. Explanation of using 2 in expected retries in testWriteRetry : 3 is max attempt, which includes 1 original request and 2 retries. Updated the constants.
6. Updated java doc specific comments.

timrobertson100 commented:
Run Java PreCommit

echauchot commented:
@aalbatross please do not squash your commits until the end of the review (just after LGTM) it is very difficult to follow review commits otherwise.

echauchot commented:
@aalbatross I see that, now, your PR contains only ES related modifications, it is good. I guess you left SolrIO retry code the way it is in the master ? there is now SolRIO retry refactoring PR, right?

echauchot commented:
@timrobertson100 I agree with your arguments I'm also -0 for common code in that case

echauchot commented:
@aalbatross I like your refactoring: duplication is limited to autovalue RetryConfiguration class and RetryPredicate interface which is not a big deal. The code is more maintainable and more inline with the other IOs for other people to maintain it

aalbatross commented:
@echauchot 
Thank you for your comments.
I agree i will avoid squashing it until LGTM in future. 
Yes you're correct, there is no SolrIO refactoring in PR anymore and SolrIO retry code is in sync with master.

aalbatross commented:
@echauchot 
I updated the changes
a. added constants for test cases
b. removed fail(), could not remove try-catch block as i need to assert against cause of exception (which is IOException), so keeping it unchanged.

echauchot commented on sdks/java/io/elasticsearch-tests/elasticsearch-tests-common/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTestCommon.java:
```diff
@@ -512,4 +536,44 @@ public Void apply(Iterable<String> input) {
+    try {
+      pipeline.run();
+    } catch (Exception ex) {
+      throw ex.getCause();
```
I guess you throw ex.getCause because Junit rule expectException expects a IOException and this exception is the cause wrapped into the exception you get ?
If so it is not needed. Junit rule suports that. In general try/catch blocks should not be needed when using expectException junit rule: you should use:
- expectedException.**expectCause**(**isA**(IOException.class))
- remove the try catch.

aalbatross commented:
@echauchot 
Thanks for the suggestion, i removed try-catch now.

echauchot commented:
there is failure in the UTests in ES2. I guess it is not related with this PR. ESIO gets an IOException when communicatig with embedded server. I guess timeout or startup issue due to load of jenkins server

echauchot commented:
retest this please

echauchot commented:
UTs pass, we only have ITs and we merge. Please update conf of thread queue on you local ES and run the ITs on green light I merge the PR

aalbatross commented:
@echauchot 
The IT tests passes successfully attaching screen shots here
<img width="1680" alt="screen shot 2018-08-23 at 23 38 16" src="https://user-images.githubusercontent.com/1793534/44553709-a5fdad80-a72e-11e8-9044-f5822bf3ad01.png">
<img width="1680" alt="screen shot 2018-08-23 at 23 09 22" src="https://user-images.githubusercontent.com/1793534/44553710-a5fdad80-a72e-11e8-9687-b620090037d8.png">
Let me know if i can squash all the commits and push them here.

echauchot commented:
LGTM, Ravi, congrats on your first Beam PR ! Keep it that way :)

echauchot commented:
Ravi, can you just squash the commits into one single named:

> [BEAM-3026] Adding retrying behavior on ESIO for error http 429 (Too Many Requests)

aalbatross commented:
I updated the PR with squashed commits. 
Thanks @echauchot @timrobertson100 to get me through with my first PR :)

echauchot commented:
retest this please

echauchot commented:
retest this please
